odinmain/src/main/kotlin/me/odinmain/features/impl/dungeon/DungeonWaypoints.kt

```diff
@@ -64,21 +74,92 @@ object DungeonWaypoints : Module(
         val filled: Boolean,
         val depth: Boolean,
         val aabb: AxisAlignedBB,
-        val title: String?
+        val title: String?,
+        val secret: Boolean,
+        var clicked: Boolean = false
     )
 
     override fun onKeybind() {
         allowEdits = !allowEdits
         modMessage("Dungeon Waypoint editing ${if (allowEdits) "§aenabled" else "§cdisabled"}§r!")
     }
 
+    private val secretEntities = mutableMapOf<Int, Vec3>()
+
+    private val drops = listOf(
+        "Health Potion VIII Splash Potion", "Healing Potion 8 Splash Potion", "Healing Potion VIII Splash Potion",
+        "Decoy", "Inflatable Jerry", "Spirit Leap", "Trap", "Training Weights", "Defuse Kit", "Dungeon Chest Key", "Treasure Talisman", "Revive Stone",
+    )
+
+    fun resetSecrets() {
+        val room = DungeonUtils.currentRoom
+        for ((_, waypointsList) in DungeonWaypointConfigCLAY.waypoints.filter { waypoints -> waypoints.value.any { it.clicked } }) {
+            waypointsList.filter { it.clicked }.forEach { it.clicked = false }
+        }
+
+        if (room != null) DungeonUtils.setWaypoints(room)
+        glList = -1
+        secretEntities.clear()
+    }
+
+    init {
+        onWorldLoad {
+            resetSecrets()
+        }
+
+        onMessage(Regex("(?s).*(\\d+)/\\1 Secrets.*")) {
+            val room = DungeonUtils.currentRoom ?: return@onMessage
+            val waypoints = DungeonWaypointConfigCLAY.waypoints.getOrPut(room.room.data.name) { mutableListOf() }
+            if (waypoints.any { it.secret && !it.clicked}) {
+                for (wp in waypoints.filter { it.secret && !it.clicked }) { wp.clicked = true }
+                DungeonUtils.setWaypoints(room)
+                glList = -1
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onMetaData(event: PostEntityMetadata) {
+        val entity = mc.theWorld.getEntityByID(event.packet.entityId) ?: return
+        if (entity is EntityItem && entity.entityItem.displayName.noControlCodes.containsOneOf(drops, true) && DungeonUtils.inDungeons) {
+            secretEntities[event.packet.entityId] = entity.positionVector
+        }
+    }
+
+    @SubscribeEvent
+    fun onEntityJoinWorldEvent(event: EntityJoinWorldEvent) {
+        if (event.entity is EntityBat && DungeonUtils.inDungeons) {
+            secretEntities[event.entity.entityId] = event.entity.positionVector
+        }
+    }
+
+    @SubscribeEvent
+    fun onEntityLeaveWorld(event: EntityLeaveWorldEvent) {
+        val pos = secretEntities[event.entity.entityId] ?: return
+        if ((event.entity is EntityItem && event.entity.entityItem.displayName.noControlCodes.containsOneOf(drops, true)) || event.entity is EntityBat) {
+            clickSecret(pos, 3)
+        }
+    }
```

Comment:
these three events are so chaotic, try to either use something to merge the detection with secret chime, or atleast use the same logic as it